src/main/java/org/datadog/jenkins/plugins/datadog/tracer/JavaConfigurator.java

```diff
@@ -0,0 +1,182 @@
+    private static final String TRACER_JAR_CACHE_TTL_ENV_VAR = "DATADOG_JENKINS_PLUGIN_TRACER_JAR_CACHE_TTL_MINUTES";
+    private static final int DEFAULT_TRACER_JAR_CACHE_TTL_MINUTES = 60 * 12;
+    private static final int TRACER_DOWNLOAD_TIMEOUT_MILLIS = 60_000;
+    private static final String DATADOG_DISTRIBUTION_HOST = "dtdg.co";
+    private static final String MAVEN_CENTRAL_HOST = "repo1.maven.org";
+    private static final String DATADOG_AGENT_MAVEN_DISTRIBUTION_PATH = "/maven2/com/datadoghq/dd-java-agent/";
+
+    private final HttpClient httpClient = new HttpClient(TRACER_DOWNLOAD_TIMEOUT_MILLIS);
+
+    @Override
+    public Map<String, String> configure(DatadogTracerJobProperty<?> tracerConfig, Node node, FilePath workspacePath, Map<String, String> envs) throws Exception {
+        FilePath tracerFile = downloadTracer(tracerConfig, workspacePath);
+        return getEnvVariables(tracerConfig, node, tracerFile, envs);
+    }
+
+    private FilePath downloadTracer(DatadogTracerJobProperty<?> tracerConfig, FilePath workspacePath) throws Exception {
+        FilePath datadogFolder = workspacePath.child(".datadog");
+        datadogFolder.mkdirs();
+
+        FilePath datadogTracerFile = datadogFolder.child(TRACER_FILE_NAME);
+        long minutesSinceModification = TimeUnit.MILLISECONDS.toMinutes(System.currentTimeMillis() - datadogTracerFile.lastModified());
+        if (minutesSinceModification < getTracerJarCacheTtlMinutes(tracerConfig)) {
+            // downloaded tracer is fresh enough
+            return datadogTracerFile.absolutize();
+        }
+
+        String tracerDistributionUrl = getTracerDistributionUrl(tracerConfig);
+        httpClient.getBinary(tracerDistributionUrl, Collections.emptyMap(), is -> {
+            try {
+                datadogTracerFile.copyFrom(is);
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+        });
+
+        return datadogTracerFile.absolutize();
+    }
+
+    private int getTracerJarCacheTtlMinutes(DatadogTracerJobProperty<?> tracerConfig) {
+        String ttl = getEnvVariable(tracerConfig, TRACER_JAR_CACHE_TTL_ENV_VAR);
+        if (ttl != null) {
+            try {
+                return Integer.parseInt(ttl);
+            } catch (Exception e) {
+                // ignored
+            }
+        }
+        return DEFAULT_TRACER_JAR_CACHE_TTL_MINUTES;
+    }
+
+    private String getTracerDistributionUrl(DatadogTracerJobProperty<?> tracerConfig) {
+        String distributionUrl = getEnvVariable(tracerConfig, TRACER_DISTRIBUTION_URL_ENV_VAR);
+        if (distributionUrl != null) {
+            validateUserSuppliedTracerUrl(distributionUrl);
+            return distributionUrl;
+        } else {
+            return DEFAULT_TRACER_DISTRIBUTION_URL;
+        }
+    }
+
+    private void validateUserSuppliedTracerUrl(String distributionUrl) {
```

Comment:
While it looks like we validate the url, is there some way for the configurator to confirm that this is a validated copy of the tracer?  Main thing guarding against here is a compromise of maven.  I notice that there are some degree of shasum checks later on - does that apply here?

Reply:
Tracer available in Maven Central is signed with Datadog's PGP key. I have added signature verification for cases when we download it from Maven.